lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.c

```diff
@@ -511,21 +523,57 @@ status_code_e s1ap_mme_handle_s1_setup_request(s1ap_state_t* state,
         S1ap_TimeToWait_v20s);
     increment_counter("s1_setup", 1, 2, "result", "failure", "cause",
                       "invalid_state");
-    // Check if the UE counters for eNB are equal.
-    // If not, the eNB will never switch to INIT state, particularly in
-    // stateless mode.
-    // Exit the process so that health checker can clean-up all Redis
-    // state and restart all stateless services.
-    AssertFatal(
-        enb_association->nb_ue_associated ==
-            enb_association->ue_id_coll.num_elements,
-        "Num UEs associated with eNB (%u) is more than the UEs with valid "
-        "mme_ue_s1ap_id (%zu). This is a deadlock state potentially caused by "
-        "misbehaving eNB; restarting MME. In stateless mode, health management "
-        "service will eventually detect multiple MME restarts due to this "
-        "deadlock state and force sctpd and hence all services to restart.",
-        enb_association->nb_ue_associated,
-        enb_association->ue_id_coll.num_elements);
+    /* Check if the UE counters for eNB are equal.
+     * If not, the eNB will never switch to INIT state, particularly in
+     * stateless mode.
+     * UE state at s1ap task is created on reception of initial ue message
+     * Hash list, ue_id_coll is updated after mme_app_task assigns and provides
+     * mme_ue_s1ap_id to s1ap task
+     * s1ap task shall clear this UE state if mme_app task has not yet provided
+     * mme_ue_s1ap_id
+     */
+    if (enb_association->ue_id_coll.num_elements == 0) {
+      unsigned int i = 0;
+      unsigned int num_elements = 0;
+      bool unlock = true;
+      hash_table_ts_t* hashtblP = get_s1ap_ue_state();
+      hash_node_t *node = NULL, *oldnode = NULL;
+      if (!hashtblP) {
+        OAILOG_ERROR(LOG_S1AP, "No UEs found in comp_s1ap_id hash list");
+        OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+      }
+      while ((num_elements < hashtblP->num_elements) && (i < hashtblP->size)) {
+        pthread_mutex_lock(&hashtblP->lock_nodes[i]);
+        if (hashtblP->nodes[i] != NULL) {
+          node = hashtblP->nodes[i];
+        }
+        while (node) {
+          num_elements++;
+          oldnode = node;
+          node = node->next;
+          if (oldnode->data) {
+            unlock = false;
+            pthread_mutex_unlock(&hashtblP->lock_nodes[i]);
+            s1ap_remove_ue(state, node->data);
```

Comment:
Considering a multi-eNB setup, we need to check if this UE was connected to the particular eNB which is resetting

Reply:
Yes I verified the changes for multi eNB scenario using test case, test_multi_enb_multi_ue.py. The changes are working fine

Reply:
Did you check for the scenario where:
1. Two eNBs connect to MME, with some UEs per eNB
2. One eNB sends SCTP reset while non-zero UEs are connected and the metrics for that eNB are out of sync.